### PR TITLE
Apply some fixes from my various branches

### PR DESCRIPTION
**Pull request by zathras-crypto**
This PR cherry picks fixes from a couple of my branches to be pushed into omnicore/0.0.10

### CODE REVIEW

**Comment by dexX7** [comment]
Thanks. I addressed some smaller points in-line.

Are you comfortable with including the changes for reindexing?

In particular, removing the reindex checks around `mastercore_handler_disc_begin()`, `mastercore_handler_disc_end()`, and replacing `new x_xxxdb(..., false)` by `new x_xxxdb(..., fReindex)`, during the database initialization in `mastercore_init()`?

**Comment by zathras-crypto** [comment]
> Are you comfortable with including the changes for reindexing?

Yep, though I'm thinking for thoroughness we should also wipe the persistence directory, we're cleaning the state so that would seem cleaner.

**Comment by dexX7** [comment]
> Well the thing is, a failure in msc_file_load() (eg a failed hash) is not necessarily a failure to load persistence.

Can you elaborate please? It shows there is something very wrong with the persisted data, and it triggers a reparse.

> Yep, though I'm thinking for thoroughness we should also wipe the persistence directory.

I agree. In this case: do we still need an option to "wipe" a database, when creating the DB object, at all?

**Comment by zathras-crypto** [comment]
> Can you elaborate please?

Sure, let's take a look at ```load_most_relevant_state()``` and the use of ```msc_file_load()```.  We have this chunk of code:
```
while (NULL != curTip && persistedBlocks.size() > 0) {
    if (persistedBlocks.find(spBlockIndex->GetBlockHash()) != persistedBlocks.end()) {
        int success = -1;
        for (int i = 0; i < NUM_FILETYPES; ++i) {
            boost::filesystem::path path = MPPersistencePath / strprintf("%s-%s.dat", statePrefix[i], curTip->GetBlockHash().ToString());
            const std::string strFile = path.string();
            success = msc_file_load(strFile, i, true);
            if (success < 0) break;
        }
        if (success >= 0) {
            res = curTip->nHeight;
            break;
        }
        // remove this from the persistedBlock Set
        persistedBlocks.erase(spBlockIndex->GetBlockHash());
    }
    // go to the previous block
    if (0 > _my_sps->popBlock(curTip->GetBlockHash())) {
        // trigger a full reparse, if the levelDB cannot roll back
        return -1;
    }
    curTip = curTip->pprev;
    if (curTip != NULL) _my_sps->setWatermark(curTip->GetBlockHash());
}
```

So you can see that if ```msc_file_load()``` fails, ```success``` will be a negative return code causing a ```break``` out of the ```for```.  The block is then removed from the ```persistedBlocks``` set and the SP database rolled back one block, following which ```msc_file_load()``` is called again.

For example calling ```msc_file_load()``` on block 300,000 returns a negative return code, so the SP database is rolled back and ```msc_file_load()``` is called again, this time on block 299,999.  Rinse repeat.

TL:DR; ```msc_file_load()``` can be called numerous times - a failure in one calling does not mean a complete failure of persistence, it means a failure to load state *for that block*.

Hope that makes sense bud :)

**Comment by zathras-crypto** [comment]
> Edit: to clarify: it's not 100 % clear to me, how you classify "failure". If the state is rolled back, due to a reorg for example, then I wouldn't consider it as "failure to load persistence".

I would.  We keep 50 blocks worth of relevant state (```MAX_STATE_HISTORY```) - if we're not able to load one of those during a reorg that I see that as a "failure to load persistence".

Sure I'll agree that if we reorg more than 50 blocks then that's not a failure to load persistence because we didn't have it stored in the first place, but I'd argue if the Bitcoin blockchain was seeing 50 block reorgs at this stage in the game we're probably all ****ed!

**Comment by zathras-crypto** [comment]
> I agree. In this case: do we still need an option to "wipe" a database, when creating the DB object, at all?

Well it depends how we come at it.  When we 'create' the DB object if a DB exists at that path it will be used, so one way or the other it needs to be wiped.  We could just use the code already there for ```-startclean``` and say if ```fReIndex``` is true then that code is also executed which would cover all DB's and the persistence directory.

**Comment by dexX7** [comment]
> TL:DR; ...

Argh, you're right. I haven't considered this is nested. I think in the current form this could be confusing, but I also have no idea how to tackle it. So let's maybe take another look at it later.

I'm currently building and then testing (again) a merged version of this PR. Is this ready to go so far?

**Comment by zathras-crypto** [comment]
> I'm currently building and then testing (again) a merged version of this PR. Is this ready to go so far?

I was going to loop around and take a quick second look at the abort stuff, as it looks like I was using it wrong (getting the "internal error" message).

**Comment by dexX7** [comment]
> I was going to loop around and take a quick second look at the abort stuff

Ah, okay.

**Comment by zathras-crypto** [comment]
Doing it now mate, won't be more than a couple of minutes - the duplicated function (```AbortOmniNode```) is not needed.

**Comment by dexX7** [comment]
> the duplicated function (AbortOmniNode) is not needed

Yeah, seems reasonable. I favor for those smallish things as rule of thumb: if the same part is used more than twice, it usually provides a benefit to extract it into a seperated function/module. In this case, well.. haha, maybe even something like:

```c++
bool mastercore::AbortOmniNode(const std::string &strMessage) {
    return AbortNode(strMessage, strMessage);
}
```

But with only one use at the moment, 8c72e08 seems reasonable. :)

**Comment by zathras-crypto** [comment]
> But with only one use at the moment

FYI the auditor also shuts down, so this may get revisited :)

**Comment by dexX7** [comment]
Yup, it was added to this branch here via this commit: https://github.com/zathras-crypto/mastercore/commit/fd6cd487f1aec9dac2f8b0c9e2930c5f10bd5d9b

**Comment by dexX7** [comment]
Mainnet: PASSED
- http://bitwatch.co/uploads/omnicore-pr-28-13/mainnet/

Regtests: 5 failures
- http://bitwatch.co/uploads/omnicore-pr-28-13/regtest/

1 is related to an output amount check (good!)
2 are related to an improvement of an error message (good!)
2 are related to `trade_MP`:
```
JsonRPCStatusException: value is type str, expected int
```

I'm not sure, where the last one is coming from, but I'll take a look at it now.

**Comment by zathras-crypto** [comment]
```trade_MP``` is now just an alias for ```sendtrade_OMNI``` so perhaps I missed something in that call?

**Comment by dexX7** [comment]
@zathras-crypto: amount/property identifier are swapped, and it looks like the version of `trade_MP` is pretty old?

See:
https://github.com/OmniLayer/omnicore/blob/omnicore-0.0.10/src/mastercore_rpc.cpp#L943-L1023
https://github.com/zathras-crypto/mastercore/blob/0.0.10-Z-ClassC/src/omnicore_rpctx.cpp#L757-L830

E.g. `"Invalid action (1,2,3,4 only)"` vs. `"Invalid action (1, 2, 3, 4 only)"`

Those changes were merged into 0.0.9.0 in December: https://github.com/mastercoin-MSC/mastercore/commit/d2d714f0ce1aa3b825ec36a758277c253cc00203

No objections here to swap amount/propertyid (it matches the spec!), but it still makes me wonder a bit, and the logic is now probably a bit outdated.

**Comment by zathras-crypto** [comment]
Ahh I see I see - nice catch.  Again I think probably the difference on collaborative development (I didn't do the original ```trade_MP``` but did write the newer ```sendtrade_OMNI```).

I'm comfortable with either option, though I find having field order in the RPC as per field order in the packet 'cleaner'.

**Comment by dexX7** [comment]
Did you recreate the commands in `omnicore_rpctx` or copied them over?

**Comment by zathras-crypto** [comment]
> Did you recreate the commands in omnicore_rpctx or copied them over?

IIRC I created them fresh.

**Comment by dexX7** [comment]
Ah sorry, the only moved/replaced commands were `send_MP`, `sendtoowners_MP` and `trade_MP`, and the other two are fine.

**Comment by dexX7** [comment]
Re: `trades_MP`:

Here is a mixed version: https://gist.github.com/dexX7/a5f815697af491047784

Tests ongonig. Fortunally I have more meta DEx tests besides the test plan, which also cover the cancel-actions, and it looks good so far. :)

Feel free to edit as you like, once the PR is updated, and also passes the tests, it's basically ready.

**Comment by zathras-crypto** [comment]
> Here is a mixed version

**Comment by dexX7** [comment]
Yes, the cancel-everything logic has different requirements for the properties. In particular, the following checks should not be applied:
```
CMPSPInfo::Entry spForSale;
if (false == _my_sps->getSP(propertyIdForSale, spForSale)) throw JSONRPCError(RPC_INVALID_PARAMETER, "Property for sale does not exist");
CMPSPInfo::Entry spDesired;
if (false == _my_sps->getSP(propertyIdDesired, spDesired)) throw JSONRPCError(RPC_INVALID_PARAMETER, "Property desired does not exist");

if (isTestEcosystemProperty(propertyIdForSale) != isTestEcosystemProperty(propertyIdDesired)) throw JSONRPCError(RPC_INVALID_PARAMETER, "Property fpr sale and property desired must be in the same ecosystem");
if (propertyIdForSale == propertyIdDesired) throw JSONRPCError(RPC_INVALID_PARAMETER, "Property for sale and property desired must be different");
```

**Comment by zathras-crypto** [comment]
> Yes, the cancel-everything logic has different requirements for the properties

Ah, I see thanks

**Comment by dexX7** [comment]
Ah, if you adopt the patch from above, `#include "mastercore_tx.h"` needs to be added, to get access to `CMPTransaction::ACTION`.

**Comment by zathras-crypto** [comment]
Thanks for the patch - I did it slightly differently to maintain the 'flow' of approach but if this isn't suitable just let me know and I'll drop it out and replace with your patch.
